### ISSUE TITLE
[Feature]  Attention-FFN Disaggregation

### ISSUE THREAD

**Issue by D-Wythe** [LPhgh/sglang#1]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Motivation

Large-scale MoE inference is often bottlenecked by conflicting resource needs. Specifically, during the decoding phase, attention layers are typically memory-bound due to KV cache operations, while the FFN/expert layers remain heavily compute-bound. A single deployment strategy cannot optimize both effectively.

A powerful architecture, Attention-FFN Disaggregation (AFD), has emerged as a key solution to this challenge. The core principle of AFD is to physically separate the attention and FFN/expert layers onto distinct sets of GPUs. This architectural disaggregation allows each component to scale independently and adopt the most suitable parallelism strategy.

To bring these benefits to SGLang, we suggest an AFD implementation that decouples Attention and FFN/experts for independent scaling. This would be achieved by pipelining their communication using a mix of data and expert parallelism. The roadmap can be found in https://github.com/sgl-project/sglang/issues/8360.

To demonstrate the feasibility of our proposed architecture, we have implemented a basic proof-of-concept (PoC), available at [SGlang-POC](https://github.com/LPhgh/sglang/tree/afd). This prototype validates the core ideas of disaggregation and pipelining, using simple communication backends such as ZMQ and StepMesh.

It is important to note that this is a minimal implementation focused solely on illustrating the basic data flow. As such, many production-critical capabilities, such as error resilience, performance optimization and CUDA graph integration are currently out of scope. We view these enhancements as future work, and intend to evolve them in collaboration with the community, aligning on design directions and practical needs before moving toward a more robust and scalable system.

---

We compare the latency of our AFD demo, with the latency w/o AFD. As shown, TTFT is slightly higher while ITL shows no difference. This is because we use 4k input length, resulting in more communication overhead at prefill stage of AFD.

```
python3 -m sglang.bench_serving --port 30000 --backend sglang --dataset-name random-ids --random-input-len 4096 --random-output-len 100 --random-range-ratio 1 --num-prompt 1
```

|                                      | Mean TTFT (ms) | Mean ITL (ms) |
|--------------------------------------|----------------|---------------|
| w/o AFD <br />(1 GPU for the whole model) | 48.17          | 34.32         |
| AFD*<br />(1 GPU for A and 1 GPU for F) | 50.64          | 34.97         |

> To simulate inter-node StepMesh communication, we bind different RDMA NICs for A and F, respectively. 

We are planning to test inter-node throughput performance at larger scale to further confirm the advantage of AFD.

---

[1] Zhu R, Jiang Z, Jin C, et al. MegaScale-Infer: Serving Mixture-of-Experts at Scale with Disaggregated Expert Parallelism[J]. arXiv preprint arXiv:2504.02263, 2025.

[2] Wang B, Wang B, Wan C, et al. Step-3 is Large yet Affordable: Model-system Co-design for Cost-effective Decoding[J]. arXiv preprint arXiv:2507.19427, 2025.

[3] https://github.com/stepfun-ai/StepMesh

### Related resources

# Proposed Change (The proposed change of the RFC)

__Schedule Synchronization__:  Continuously synchronize the Schedule batch between attention instance and FFN instance to ensure their execution remains aligned. (Note: This part is certainly debatable. We proposed it this way to maximize the reuse of existing SGLang features, but we are very open to alternative approaches)

![Image](https://github.com/user-attachments/assets/2be765c2-5bbf-4626-b950-36c2e4332324)

__Role-Specific Engine Deployment__: Partition the model's computation graph and deploy the attention and FFN stages onto separate SGLang engine instances. Each instance is then responsible for loading only the resources relevant to its role on-demand.

![Image](https://github.com/user-attachments/assets/1f91b022-85d0-4103-a5cf-3158ecf29ab9)

__Pipelining and overlap__: Enable a cross-stage pipeline by splitting batches into micro-batches. This allows for overlapping the computation of one stage with the communication of another, orchestrated through the abstract interface defined above.

![Image](https://github.com/user-attachments/assets/d53d3e2f-a1a6-4be8-92a6-f8be1e59c1fd)

__Abstract Communication Interface__: Define an abstract collective communication interface for handling tensor delivery between Attention and FFN, designed to be agnostic to the underlying transport (be it stepmesh or mooncake) and the execution hardware (CPU or GPU).

![Image](https://github.com/user-attachments/assets/8a201bb2-637b-4dee-84b6-4f0f2991f1fd)

### Milestone 0: Proof of Concept

- [x] **Intra-Node Disaggregation**: Partitioned the Qwen3-Moe computation graph, enabling the deployment of its attention and FFN stages as dedicated SGLang instances within a single node. 
- [x] **A-F scheduler synchronization**: Implemented a event_loop_normal variant, where A accepts generation requests and forwards them to F.
- [x] **Cross-Instance Communication**: Delivery tensor between A/F instances through an initial integration of StepMesh and ZMQ communication backends.
- [x] **Mirco-Batch and Pipelined Execution**: Implemented a cross-stage pipeline with a hardcoded batch split (m=2 or m=3) to enable basic computation-communication overlap.
- [ ] **Inter-Node Disaggregation**:  Support the deployment of  attention and FFN stages as dedicated SGLang instances within distributed node.
- [ ] **Performance Test Report (Derived from Proof of Concept)**: More evaluations for precision and throughput performance.

### Milestone 1: Basic Functionality

- [ ] **Unified instances management**: Design a formalized schema for instances topology, add support to static layout for distributed deployment.
- [ ] **Generalized Batch Overlap (N-Batch Overlap)**: Evolve the current two-batch overlap mechanism into a configurable N-batch overlap.
- [ ] **Abstract Communication Interface**: Define an abstract collective communication interface for handling tensor delivery between Attention and FFN, Initial support will be provided via StepMesh as a lightweight backend.
- [ ] **Improved A-F scheduler synchronization**: Architecting a formal, high-performance scheduler synchronization to enhance accuracy and performance in activation-forward scheduling.
- [ ] **Expand Model Support**: Add compatibility for more models like Step-3 and DeepSeek.

### Milestone 2: Optimizations

- [ ] **More Efficient MoE Communication**: Enable sparse data exchange through token-based expert routing using router logits and expert mapping.
- [ ] **CUDA Graph Integration**: Add support for CUDA Graph to improve kernel launch efficiency.
- [ ] **Standardized Benchmarking**: Establish official throughput and latency baselines for consistent performance evaluation.

### Milestone 3: Enhancements

- [ ] **Full parallelism compatibility**: Add compatibility with PP, Attention DP. etc
- [ ] **Lightweight FFN Engine Core**: Create a minimal SGLang engine core for FFN-only instances, eliminating non-essential components (e.g., KV cache manager).
- [ ] **Better microbatch splitting policy**:  Deciding whether and how to split batches for more balanced workload and better throughput performance. 
- [ ] **Overlap A-F scheduler**: Add overlap support for the AFD event loop.
- [ ] **Full disaggregation**: Add compatibility with PD disaggregation.

# Any Other Things.

We believe this direction holds promise, and progress will depend on collaboration with the community. We welcome input, discussion, and contributions from those interested in exploring disaggregated inference architectures. Practical feedback and shared insights will be key to shaping its evolution.
